lib/SILOptimizer/Utils/Generics.cpp

```diff
@@ -171,6 +273,198 @@ createSpecializedType(CanSILFunctionType SubstFTy, SILModule &M) const {
+
+  auto OrigGenericSig = OrigF->getLoweredFunctionType()->getGenericSignature();
+  auto OrigGenericEnv = OrigF->getGenericEnvironment();
+  SpecializedGenericSig = OrigGenericSig;
+  auto ForwardingSubs = OrigGenericEnv->getForwardingSubstitutions();
+  auto ForwardingInterfaceSubsMap =
+      OrigGenericSig->getSubstitutionMap(ForwardingSubs);
+  // Map archetypes of the original function to the contextual types
+  // of the specialized function.
+  SubstitutionMap ClonerArchetypeToConcreteMap;
+  SubstitutionMap CallerArchetypeToConcreteMap;
+
+  SmallVector<RequirementRepr, 4> RequirementReprs;
+  convertRequirements(Requirements, RequirementReprs);
+
+  for (auto &Req : Requirements) {
+    if (Req.getKind() == RequirementKind::SameType) {
+      // Remember that a given generic parameter is mapped
+      // to a concrete type.
+      CallerInterfaceSubs.addSubstitution(
+          dyn_cast<SubstitutableType>(Req.getFirstType()->getCanonicalType()),
+          Req.getSecondType()->getCanonicalType());
+
+      // Remember how the original interace type is mapped to a concrete type.
+      ClonerArchetypeToConcreteMap.addSubstitution(
+          dyn_cast<SubstitutableType>(
+              OrigGenericEnv->mapTypeIntoContext(SM, Req.getFirstType())
+                  ->getCanonicalType()),
+          Req.getSecondType()->getCanonicalType());
+
+      // Remember how the original interace type is mapped to a concrete type.
+      CallerArchetypeToConcreteMap.addSubstitution(
+          dyn_cast<SubstitutableType>(
+              OrigGenericEnv->mapTypeIntoContext(SM, Req.getFirstType())
+                  ->getCanonicalType()),
+          Req.getSecondType()->getCanonicalType());
+    }
+  }
+
+  // Perform some sanity checks for the requirements
+  checkSpecializationRequirements(RequirementReprs);
+
+  std::tie(SpecializedGenericEnv, SpecializedGenericSig) =
+      getSignatureWithRequirements(OrigGenericSig, OrigGenericEnv,
+                                   RequirementReprs, M);
+
+  for (auto Req : SpecializedGenericSig->getRequirements()) {
+    // Remember how the original contextual type is represented in
+    // the specialized function.
+    if (Req.getKind() == RequirementKind::Layout) {
+      ClonerArchetypeToConcreteMap.addSubstitution(
+          dyn_cast<SubstitutableType>(
+              OrigGenericEnv->mapTypeIntoContext(SM, Req.getFirstType())
+                  ->getCanonicalType()),
+          SpecializedGenericEnv->mapTypeIntoContext(SM, Req.getFirstType()));
+    }
+  }
+
+  // Substitute into forwarding substitutions to get the cloner substitutions
+  // and substitutions to be used by the caller for calling the specialized
+  // function.
+  SmallVector<Substitution, 4> ClonerSubs;
+  SmallVector<Substitution, 4> CallerSubs;
+  for (auto Sub : ForwardingSubs) {
+    auto NewClonerSub = Sub.subst(
+        SM, QueryTypeSubstitutionMap{ClonerArchetypeToConcreteMap.getMap()},
+        LookUpConformanceInModule(SM));
```

Comment:
I looked at the referenced SR and indeed the stack trace looks very similar to what I was observing.